### ISSUE TITLE
Location way_id parameter processing

### ISSUE THREAD

**Issue by alexey-gl** [valhalla/valhalla#3729]
Hello!

> Location way_id
> (optional) OpenStreetMap identification number for a polyline [way](http://wiki.openstreetmap.org/wiki/Way). The way ID may be used to assist finding the correct routing location at the specified latitude, longitude. This is not currently implemented.

I'm going to use way_id parameter and implement it's processing. Does smb have started such work? Kindly ask give me some advices in that way.

**Comment by kevinkreiser** [comment]
the work is not started but i would be happy to give you a rough outline of what needs to be done and where i the code. does that sounds like a plan?

**Comment by kevinkreiser** [comment]
@alexey-gl that is correct but i think before we start making changes we need to do a little design work. so lets maybe ask and answer some questions about how it should function. specifically i am thinking about one big  question right now which is:

1. is the `way_id` a hard or soft filter? that is, must `loki::search` return an edge with matching `way_id` or can it ignore the `way_id` in situations where no such edges are found near by?

from a functinality standpoint, i prefer the soft filter similar to how the `heading` filter works. basically we can favor edges who have a way_id that matches the input but if there aren't any near by we can fall back to ignoring it. the problem with this option is its more difficult to implement when you consider the radius defaults to 0 which means give me the closest thing. the heading filter gets around this by storing filtered edges whose heading didnt match in a separate set of edges and then pulling from them when nothing better was found for computing a route. such a strategy would not work in this case because the intent is, if you see the way_id  you should use it unless its 20km away or something?

the alternative hard filter which would ignore all other edges who dont have a matching way_id is much simpler to implement than the soft filter but there is a big drawback. in this case the performance will be quite poor when the person asks for a way_id that is no where near the input location. this would cause the search to spiral out until it finally finds an edge with that way_id or eventually non at all (hitting the search cutoff of 30km default).

i think to decide which one of these we should go with we should consider the use-cases for specifying a way_id. this should make it clear which approach covers more or less of the use-cases we come up with. so my question to you, do you have a specific use-case in mind? which approach had you envisioned?

**Comment by alexey-gl** [comment]
@kevinkreiser 
Thinking about this feature I assume to have a hard filter: either correlated edge were found, either route not found at all.

My current use-case: we have a long route, which was retrieved earlier. Client should be able to edit only part of the route. In my case it is important that the updated part of the route connected properly with the rest of the original route, so that route stay valid after editing (i.e. without gaps or overlaps).

At the same time I keep in mind another use-case which will require hard filter. If Way connected to some object (building or smth) which should be visited for sure. So client will mention way_id parameter to be sure route will pass this Way.

So both my cases require hard filter strategy. 

To solve performance problem I think it is possible to set a much smaller search cutoff in case way_id parameter is set.

**Comment by kevinkreiser** [comment]
Ok lets implement the hard filter then. I would suggst you do the following:

1. add a new filter here: https://github.com/valhalla/valhalla/blob/master/proto/common.proto#L34 that looks something like: `repeated uint64_t allow_way_ids = 7;` that way we can whitelist specific way_ids in the filter
2. add parsing for this somehwere near here to add to a location object: https://github.com/valhalla/valhalla/blob/master/src/worker.cc#L365  here you just put them into the protobuf repeated field you made in the previous step
3. then you need to be able to convert them to the native c++ object from protobuf here: https://github.com/valhalla/valhalla/blob/master/valhalla/baldr/pathlocation.h#L227 and here: https://github.com/valhalla/valhalla/blob/master/valhalla/baldr/pathlocation.h#L132 please note that when you store it in the native non protobuf object it shouldnt be an array but rather an unordered_set
4. finally you come over to here: https://github.com/valhalla/valhalla/blob/master/src/loki/search.cc#L38 and add the check, if the unordered_set is empty the check just ignores it but if it is not empty then if the way_id is not in the set the edge is filtered. what sucks about this is getting the way_id requires getting the edge info which isnt currently passed to this function and that is on purpose because its more expensive to go get that data.. perhaps we should take a little pain to check if the filter requires us check the way id (ie if the unordered set is not empty) and only then go fetch the edge info. we can worry about that further along in the implementation..
5. after that you should be updating some tests to make sure it works as it should. we have the old tests/search.cc test but its certainly better to do in a gurka test. ill look around and suggest one at a later point

**Comment by alexey-gl** [comment]
@kevinkreiser 
Thank you for extensive comments. I will go dive into it.
